src/irmin-pack/unix/gc_worker.ml

```diff
@@ -265,21 +269,24 @@ module Make (Args : Gc_args.S) = struct
             Sparse.Ao.append_seq_exn prefix ~off str)
           live_entries
       in
-      (* Step 5.2. Update the parent commits to be dangling.
-         Reopen the new prefix, this time in write-only as we have to modify data inside the file. *)
-      stats :=
-        Gc_stats.Worker.finish_current_step !stats
-          "prefix: rewrite commit parents";
-      let prefix = Sparse.Wo.open_wo ~mapping ~data |> Errs.raise_if_error in
-      Errors.finalise_exn (fun _outcome ->
-          Sparse.Wo.fsync prefix
-          >>= (fun _ -> Sparse.Wo.close prefix)
-          |> Errs.log_if_error "GC: Close prefix after parent rewrite")
-      @@ fun () ->
-      let write_exn = Sparse.Wo.write_exn prefix in
-      List.iter
-        (fun key -> transfer_parent_commit_exn ~write_exn key)
-        (Commit_value.parents commit)
+      (* Step 5.2.
+         Update the parent commits to be dangling, but only if we are deleting
+         the data: reopen the new prefix, this time in write-only as we have to
+         modify data inside the file. *)
+      if Fm.gc_behaviour fm = `Delete then (
```

Comment:
This is an unexpected guard to me. Even if we are archiving I think we want to mark the parents as dangling since the data is not in the prefix part of the data.

Reply:
The removal of that guard causes data in the lower to fail to be found, more specifically in `pack_store.ml:214`
```ocaml
  (* This function assumes magic is written at hash_size + 1 for every
     object. *)
  let gced buf =
    let kind = Pack_value.Kind.of_magic_exn (Bytes.get buf Hash.hash_size) in
    kind = Pack_value.Kind.Dangling_parent_commit

  let pack_file_contains_key t k =
    let off, _, volume_identifier = get_location t k in
    let len = Hash.hash_size + 1 in
    let buf = Bytes.create len in
    let (_volume : Lower.volume_identifier option) =
      Dispatcher.read_exn t.dispatcher ~off ~len ?volume_identifier buf
    in
    if gced buf then false
   (* ... *)
```
`gced` will return `true` and the data will be considered not found.

I don't have a firm handle on the chain of events and their reasons, so maybe the fix should be somewhere else in my code :thinking:

Reply:
Ah, I see. My hunch is maybe it needs to be fixed somewhere in the reading code and not here. Let me pull your branch and play with it a little.

Reply:
Here is a patch that moves the fix for this issue. https://github.com/mirage/irmin/compare/gc-archival...metanivek:dangling_parent_read

I'm not totally sure the impact of letting pack value decode the dangling parents commits, but that kind is used to know when to not propagate the volume identifier when reading data from a volume. https://github.com/mirage/irmin/blob/2b975d0a9deaa4e4f493c5c4c188bfb7c43ed183/src/irmin-pack/unix/pack_store.ml#L287-L290

But, perhaps there is a different solution altogether.

Would love @art-w's eyes, brain, and heart on this problem, too.

Reply:
Thank you for playing with it! I'll apply your patch for now, we can come back to it if need be.